### PR TITLE
Remove currIdx from DataChunkState

### PR DESCRIPTION
**Pull request by andyfengHKU**
We take another step in solving #744. This time we remove currIdx and replace with a FactorizationState.

### CODE REVIEW

File: src/include/common/data_chunk/data_chunk_state.h [hunk 1]
```diff
@@ -27,21 +28,26 @@ class DataChunkState {
...
+
+    inline uint64_t getNumSelectedValues() const { return selVector->selectedSize; }
 
 public:
-    // The currIdx is >= 0 when vectors are flattened and -1 if the vectors are unflat.
-    int64_t currIdx;
+    std::shared_ptr<SelectionVector> selVector;
+
+private:
+    FactorizationStateType factorizationState = FactorizationStateType::UNFLAT;
```
Comment (ray6080) [inline src/include/common/data_chunk/data_chunk_state.h:43]: Initialize this in constructor.

File: src/include/common/data_chunk/data_chunk_state.h [hunk 2]
```diff
@@ -27,21 +28,26 @@ class DataChunkState {
...
+    inline bool isFlat() const { return factorizationState == FactorizationStateType::FLAT; }
+    inline void setToFlat() { factorizationState = FactorizationStateType::FLAT; }
+    inline void setToUnflat() { factorizationState = FactorizationStateType::UNFLAT; }
+
+    inline uint64_t getNumSelectedValues() const { return selVector->selectedSize; }
 
 public:
-    // The currIdx is >= 0 when vectors are flattened and -1 if the vectors are unflat.
-    int64_t currIdx;
+    std::shared_ptr<SelectionVector> selVector;
```
Comment (ray6080) [inline src/include/common/data_chunk/data_chunk_state.h:40]: Also move this field `selVector` to private?

File: test/transaction/transaction_test.cpp
Comment (ray6080) [inline test/transaction/transaction_test.cpp:1]: Why remove this test?

File: src/include/common/data_chunk/data_chunk_state.h (hunk 1 above)
Comment (ray6080) [inline src/include/common/data_chunk/data_chunk_state.h:43]: Not related to this PR. but a suggestion is we can simplify `FactorizationStateType` to `FStateType`, which reads easier.

File: src/processor/operator/hash_join/hash_join_probe.cpp [hunk 3]
```diff
@@ -110,6 +110,13 @@ uint64_t HashJoinProbe::getLeftJoinResult() {
         for (auto& vector : vectorsToReadInto) {
             vector->setAsSingleNullEntry();
         }
+        // TODO(Xiyang): We have a bug in LEFT JOIN which should not discard NULL keys. To be more
+        // clear, NULL keys should only be discarded for probe but should not reflect on the vector.
+        // The following for loop is a temporary hack.
```
Comment (ray6080) [inline src/processor/operator/hash_join/hash_join_probe.cpp:115]: Maybe open an issue one this if we are not going to fix this immediately?

**Review by ray6080 — APPROVED** [review]

File: src/include/common/data_chunk/data_chunk_state.h (hunk 2 above)
Comment (andyfengHKU) [reply src/include/common/data_chunk/data_chunk_state.h:40]: This involves a lot of change so should be done separately.

File: test/transaction/transaction_test.cpp
Comment (andyfengHKU) [reply test/transaction/transaction_test.cpp:1]: Because it is using currIdx directly. I found it hard to maintain.

**Pull request merged**